parser-typechecker/src/Unison/Codebase/Path.hs

```diff
@@ -195,22 +193,27 @@ parseHQSplit s = case parseHQSplit' s of
   Left e -> Left e
 
 parseHQSplit' :: String -> Either String HQSplit'
-parseHQSplit' s =
-  case Text.breakOn "#" $ Text.pack s of
-    ("","") -> error $ "encountered empty string parsing '" <> s <> "'"
-    ("", _) -> Left "Sorry, you can't use a hash-only reference here."
-    (n, "") -> do
-      (p, rem) <- parsePathImpl' (Text.unpack n)
-      seg <- definitionNameSegment rem
-      pure (p, HQ'.NameOnly seg)
-    (n, sh) -> do
-      (p, rem) <- parsePathImpl' (Text.unpack n)
-      seg <- definitionNameSegment rem
-      maybeToRight (shError s) .
-        fmap (\sh -> (p, HQ'.HashQualified seg sh)) .
-        SH.fromText $ sh
-  where
+parseHQSplit' s = case Text.breakOn "#" $ Text.pack s of
+  ("", "") -> error $ "encountered empty string parsing '" <> s <> "'"
+  ("", _ ) -> Left "Sorry, you can't use a hash-only reference here."
+  (n , "") -> do
+    (p, rem) <- parsePath n
+    seg      <- definitionNameSegment rem
+    pure (p, HQ'.NameOnly seg)
+  (n, sh) -> do
+    (p, rem) <- parsePath n
+    seg      <- definitionNameSegment rem
+    maybeToRight (shError s)
+      . fmap (\sh -> (p, HQ'.HashQualified seg sh))
+      . SH.fromText
+      $ sh
+ where
   shError s = "couldn't parse shorthash from " <> s
+  parsePath n = do
+    x <- parsePathImpl' $ Text.unpack n
+    pure $ case x of
+      (Path' (Left e), "") | e == absoluteEmpty -> (relativeEmpty', ".")
+      x -> x
```

Comment:
The main change here is `parsePath` which treats `.` as the _object named `.`_ instead of the absolute namespace root. The thinking here is that `parseHQSplit'` should never be used to parse a path without an object name at the end.